### PR TITLE
Fix constant folding bug in jnp.arange for non-zero start values

### PR DESCRIPTION
**Pull request by yashwantbezawada**
This fixes issue #32542 where `jnp.arange` was incorrectly folding into a constant array when called with a non-zero start value.

**Problem:**
When calling `jnp.arange(1, n)`, JAX was creating a constant array instead of using the `iota` primitive. This caused compilation time to explode for large arrays because the entire array was embedded in the computation graph.

**Root cause:**
The optimization path in `_arange` only applied `lax.broadcasted_iota` when `start == 0`. For any other start value, it fell back to `array(np.arange(...))`, which creates a concrete numpy array first and then converts it to a JAX array, triggering constant folding.

**Fix:**
Extended the optimization to handle any start value:
- Use `lax.broadcasted_iota` to generate the base sequence [0, 1, 2, ...]
- When start != 0, add the offset using `lax.add`
- This ensures the arange is computed using primitives rather than folded into constants

**Testing:**
The fix correctly handles:
- Zero start values (existing behavior preserved)
- Non-zero start values (bug fix)
- Empty ranges (size <= 0)
- Float start/stop values (using ceil)

Fixes #32542

### CODE REVIEW

File: jax/_src/numpy/lax_numpy.py [hunk 1]
```diff
@@ -5989,9 +5989,16 @@ def _arange(start: ArrayLike | DimSize, stop: ArrayLike | DimSize | None = None,
...
+    if step is None and stop is not None:
+      # Use iota + offset instead of creating a constant array
+      size = int(np.ceil(stop - start))
+      if size <= 0:
+        return array([], dtype=dtype, device=out_sharding)
+      result = lax.broadcasted_iota(dtype, (size,), 0, out_sharding=out_sharding)
+      if start != 0:
+        # Add offset if start is non-zero
+        result = lax.add(result, lax.convert_element_type(start, dtype))
+      return result
```
[the remark this answers is not in the document]
Comment (yashwantbezawada) [reply jax/_src/numpy/lax_numpy.py:6008]: The concern about dtype=int with float start/stop is based on an incorrect premise.

The bot claimed `np.arange(0.1, 0.9, dtype=int)` returns `[]`, but it actually returns `[0]`:

```python
>>> import numpy as np
>>> np.arange(0.1, 0.9, dtype=int)
array([0])
```

My implementation correctly handles this:
- size = ceil(0.9 - 0.1) = 1
- iota(1, dtype=int) = [0]
- add int(0.1) = [0] + 0 = [0] ✓

Verified for multiple cases:
- arange(0.1, 0.9, dtype=int): [0] ✓
- arange(1.5, 5.7, dtype=int): [1,2,3,4,5] ✓
- arange(0.5, 2.5, dtype=int): [0,1] ✓

The implementation is correct.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Thanks for this! It looks reasonable, but we will need test coverage for the new logic.

In particular, you could add a sample to `testArangeJaxpr` to confirm that the new code produces a single call to `iota_p` with nonzero startpoint.

Also, we should add another test case covering some of the potential corner cases with the new implementation, such as the correctness of `np.ceil` in your implementation.

**Comment by yashwantbezawada** [comment]
Thanks for the feedback! I've added test coverage for both:

1. Extended `testArangeJaxpr` with a `(1, 5)` case to verify non-zero start uses `iota_p` (with an `add` for the offset)
2. Added `testArangeCeilCornerCases` covering float values, negative ranges, empty ranges, and boundary cases for the `np.ceil` logic

Let me know if you'd like any adjustments to the tests.

File: jax/_src/numpy/lax_numpy.py [hunk 2]
```diff
@@ -5989,9 +5989,16 @@ def _arange(start: ArrayLike | DimSize, stop: ArrayLike | DimSize | None = None,
       start = ceil_(start).astype(int)
     return lax.broadcasted_iota(dtype, (start,), 0, out_sharding=out_sharding)  # type: ignore[arg-type]
   else:
-    if step is None and start == 0 and stop is not None:
-      return lax.broadcasted_iota(dtype, (np.ceil(stop).astype(int),), 0,
-                                  out_sharding=out_sharding)
+    if step is None and stop is not None:
+      # Use iota + offset instead of creating a constant array
+      size = int(np.ceil(stop - start))
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:6001]: We should add test coverage for when `start` and/or `stop` are complex – I think this line would raise an error.

File: jax/_src/state/discharge.py
Comment (jakevdp) [inline jax/_src/state/discharge.py:1]: Changes in this file and others in `jax/_src/state` look unrelated. Please revert.

File: tests/lax_numpy_test.py [hunk 3]
```diff
@@ -4855,20 +4855,62 @@ def testArangeJit(self):
...
     device = jax.devices()[-1] if specify_device else None
     kwargs = {"device": device}
     jaxpr = jax.make_jaxpr(lambda: jnp.arange(*args, **kwargs))()
-    # We have 2 statements in jaxpr:
-    # [a:i32[5] = iota[dimension=0 dtype=int32 shape=(5,)],
-    #  a:i32[5] = device_put[devices=[None] srcs=[None]] b]
-    num_eqs = 2 if device is not None else 1
+
+    # Determine expected number of equations
+    start = 0 if len(args) == 1 else args[0]
```
Comment (jakevdp) [inline tests/lax_numpy_test.py]: Tests should have minimal logic. Rather than putting this logic in the test, we should write a separate test for the case where the start value is non-zero.

File: tests/lax_numpy_test.py [hunk 4]
```diff
@@ -4855,20 +4855,62 @@ def testArangeJit(self):
...
     self.assertEqual(jaxpr.jaxpr.eqns[0].primitive, lax.iota_p)
 
+    if has_add:
+      self.assertEqual(jaxpr.jaxpr.eqns[1].primitive, lax.add_p)
+
+  def testArangeCeilCornerCases(self):
+    """Test corner cases with np.ceil in arange implementation."""
+    test_cases = [
+      # Float values where ceiling affects result
+      (1.5, 5.5),   # ceil(4.0) = 4
```
Comment (jakevdp) [inline tests/lax_numpy_test.py]: The comments showing what `ceil` is in each case are probably not necessary. I think we could achieve more-or-less the same test coverage much more concisely by generating random start and stop values using `jtu.rand_default`, for `float32` and `int32` dtypes.

File: tests/lax_numpy_test.py [hunk 5]
```diff
@@ -4855,20 +4855,62 @@ def testArangeJit(self):
...
+      (0, 5),       # ceil(5) = 5
+      (1, 5),       # ceil(4) = 4
+    ]
+
+    for start, stop in test_cases:
+      with self.subTest(start=start, stop=stop):
+        jax_result = jnp.arange(start, stop)
+        np_result = np.arange(start, stop)
+        self.assertArraysEqual(jax_result, np_result)
+
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:4907]: Please add a test here for complex start/stop/step values as well. I'm not entirely sure what the semantics are in that case, but we should match what `np.arange` does.

**Review by jakevdp — COMMENTED** [review]
Thanks! A couple more changes requested below.

**Comment by yashwantbezawada** [comment]
Thanks for the thorough review! I've addressed all the feedback:

1. Added a check to skip the optimization when start or stop is complex (they now fall through to NumPy's implementation)
2. Reverted the unrelated state changes - this PR now only contains the arange fix
3. Split testArangeJaxpr into separate simpler tests without complex logic
4. Replaced the verbose test cases with randomized testing using jtu.rand_default
5. Added test coverage for complex start/stop/step arguments

All tests now match NumPy's behavior. Let me know if you'd like any other changes!

File: jax/_src/numpy/lax_numpy.py (hunk 2 above)
Comment (yashwantbezawada) [reply jax/_src/numpy/lax_numpy.py:6001]: Good catch\! I've added a check to skip the optimization when start or stop is complex (they now fall through to NumPy's implementation), along with test coverage.

File: jax/_src/state/discharge.py
Comment (yashwantbezawada) [reply jax/_src/state/discharge.py:1]: You're right, that was from a different fix I was working on. I've reverted these changes and kept this PR focused on just the arange fix.

File: tests/lax_numpy_test.py (hunk 3 above)
Comment (yashwantbezawada) [reply tests/lax_numpy_test.py]: Makes sense - I've split this into a separate test for the non-zero start case.

File: tests/lax_numpy_test.py (hunk 4 above)
Comment (yashwantbezawada) [reply tests/lax_numpy_test.py]: Good point - I've replaced this with randomized testing for more concise coverage.

File: tests/lax_numpy_test.py (hunk 5 above)
Comment (yashwantbezawada) [reply tests/lax_numpy_test.py:4907]: Will do - I've added test coverage for complex arguments.

File: tests/lax_numpy_test.py [hunk 6]
```diff
@@ -4869,6 +4869,39 @@ def testArangeJaxpr(self, args, specify_device):
...
+    # Non-zero start should produce iota + add (+ device_put if device specified)
+    num_eqs = 3 if device is not None else 2
+    self.assertEqual(len(jaxpr.jaxpr.eqns), num_eqs)
+    self.assertEqual(jaxpr.jaxpr.eqns[0].primitive, lax.iota_p)
+    self.assertEqual(jaxpr.jaxpr.eqns[1].primitive, lax.add_p)
+
+  @jtu.sample_product(dtype=[np.int32, np.float32])
+  def testArangeRandomValues(self, dtype):
+    rng = jtu.rand_default(self.rng())
+    for _ in range(20):
```
Comment (jakevdp) [inline tests/lax_numpy_test.py]: We should avoid this kind of control flow within the test, because if an iteration fails it obscures failures in subsequent iterations.

Since each test case gets its own random seed, you could do this within the parameterization:
```python
  @jtu.sample_product(
      dtype=[np.int32, np.float32],
      iteration=range(10)
  )
  def testArangeRandomValues(self, dtype, iteration):
    del iteration  # not needed: each test case gets its own random seed.
    rng = jtu.rand_default(self.rng())
    start = rng((), dtype)
    stop = rng((), dtype)
    jax_result = jnp.arange(start, stop)
    np_result = np.arange(start, stop)
    self.assertArraysEqual(jax_result, np_result)

Comment (yashwantbezawada) [reply tests/lax_numpy_test.py]: Good point - I've restructured the test to use parameterization instead of a loop. Each test case now gets its own random seed and runs independently.

**Review by jakevdp — COMMENTED** [review]
Looks good, thanks!

Last thing: can you please squash your changes into a single commit? See https://docs.jax.dev/en/latest/contributing.html#single-change-commits-and-pull-requests

**Comment by yashwantbezawada** [comment]
Done! Squashed all commits into a single commit as requested.

**Review by jakevdp — APPROVED** [review]
Thanks!

**Comment by jakevdp** [comment]
It looks like there are some test failures. You should be able to reproduce these locally for easier debugging with something like the following:
```
JAX_ENABLE_X64=1 JAX_NUM_GENERATED_CASES=20 pytest tests/lax_numpy_test.py -k testArange
```
It looks like the tolerances need to be adjusted.

**Review by jakevdp — APPROVED** [review]
Thanks!

File: tests/lax_numpy_test.py [hunk 7]
```diff
@@ -4869,6 +4869,42 @@ def testArangeJaxpr(self, args, specify_device):
...
+      dtype=[np.int32, np.float32],
+      iteration=range(10)
+  )
+  def testArangeRandomValues(self, dtype, iteration):
+    del iteration  # not needed: each test case gets its own random seed.
+    rng = jtu.rand_default(self.rng())
+    start = rng((), dtype)
+    stop = rng((), dtype)
+    jax_result = jnp.arange(start, stop)
+    np_result = np.arange(start, stop)
```
Comment (jakevdp) [inline tests/lax_numpy_test.py]: There are some test failures related to mismatched dtypes here. Can you add an explicit `dtype=dtype` to both the `jnp.arange` and `np.arange` calls?

Comment (yashwantbezawada) [reply tests/lax_numpy_test.py]: Fixed, thanks!

**Review by jakevdp — APPROVED** [review]
